File: src/ee/ad_colony/private/AdColonyBridge.swift
//
//  AdColonyBridge.swift
//  ee-x-7d8a1a4f
//
//  Created by eps on 11/20/20.
//

import AdColony
import RxSwift

private let kPrefix = "AdColonyBridge"
private let kInitialize = "\(kPrefix)Initialize"

@objc(EEAdColony)
public class AdColonyBridge: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _initializing = false
    private var _initialized = false

    public required init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
    }

    func registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { message, resolver in
            self.initialize(message)
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
    }

    func initialize(_ appId: String) -> Single<Bool> {
        return Single<Bool>.create { single in
            Thread.runOnMainThread {
                if self._initializing {
                    single(.success(false))
                    return
                }
                if self._initialized {
                    single(.success(true))
                    return
                }
                AdColony.configure(withAppID: appId, zoneIDs: [], options: nil) { _ in
                    self._initializing = false
                    self._initialized = true
                    single(.success(true))
                }
            }
            return Disposables.create()
        }
        .subscribeOn(MainScheduler())
    }
}

File: src/cpp/ee/ad_colony/private/AdColonyBridge.cpp
#include "ee/ad_colony/private/AdColonyBridge.hpp"

#include <ee/core/ILogger.hpp>
#include <ee/core/IMessageBridge.hpp>
#include <ee/core/Task.hpp>
#include <ee/core/Utils.hpp>
#include <ee/nlohmann/json.hpp>

namespace ee {
namespace ad_colony {
namespace {
const std::string kPrefix = "AdColonyBridge";
const std::string kInitialize = kPrefix + "Initialize";
} // namespace

using Self = Bridge;

Self::Bridge(IMessageBridge& bridge, ILogger& logger,
             const Destroyer& destroyer)
    : bridge_(bridge)
    , logger_(logger)
    , destroyer_(destroyer) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
}

Self::~Bridge() = default;

void Self::destroy() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    destroyer_();
}

Task<bool> Self::initialize(const std::string& appId,
                            const std::vector<std::string>& zoneIds) {
    nlohmann::json json;
    json["appId"] = appId;
    json["zoneIds"] = zoneIds;
    auto response = co_await bridge_.callAsync(kInitialize, json.dump());
    co_return core::toBool(response);
}
} // namespace ad_colony
} // namespace ee
